feat(AIP-121): require Get for mutable resources

toumorokoshi opened the pull request:
Resources that do not support Get but support mutable operations make customers unable to validate the
state of their resources for long-term synchronization use cases like drift detection or state management.

See https://github.com/pulumi/pulumi-google-native/issues/47 for an example of the issue.

toumorokoshi committed:
feat(AIP-121): require Get for mutable resources

Resources that do not support Get but support mutable
operations make customers unable to validate the
state of their resources for long-term synchronization
use cases like drift detection or state management.

See https://github.com/pulumi/pulumi-google-native/issues/47
for an example of the issue.

shwoodard commented on aip/general/0121.md:
```diff
@@ -117,6 +121,8 @@ and in turn do not increase resource management complexity.
 
 ## Changelog
 
+- **2022-12-19**: Added a section requiring a Get if Create,Update, or Delete
```
Prepend a space before, "Update".

toumorokoshi replied:
done.

shwoodard commented on aip/general/0121.md:
```diff
@@ -117,6 +121,8 @@ and in turn do not increase resource management complexity.
 
 ## Changelog
 
+- **2022-12-19**: Added a section requiring a Get if Create,Update, or Delete
```
Remove "a" in "requiring a Get".

toumorokoshi replied:
done.

shwoodard commented on aip/general/0121.md:
```diff
@@ -61,6 +61,10 @@ exposes a large number of resources with a small number of methods on each
 
+A resource that supports a [Create][], [Update][], or [Delete][] must also
+support [Get][] to enable the user to validate the success of their operation
+outside of the operation completion.
```
Is "outside of the operation completion" necessary? Could this sentence just end with, "...validate the success of their operation."?

Maybe, "...to validate the success of the mutation." Then we can change "their" to "the" so as not to anthropomorphize an http client?

toumorokoshi replied:
yeah, you don't need that tidbit per se. It's just there's other ways to "verify completion", e.g. via the status code on a request or an LRO. 

I didn't want people to see that and think an LRO can already tell you, so we don't need to do get.

toumorokoshi replied:
I'm going to use "validate the state of the resource", since that's actually what's being done here. Not so much "operation completion", which should be detected via LRO or request completion

toumorokoshi replied:
PTAL.

shwoodard commented on aip/general/0121.md:
```diff
@@ -61,6 +61,10 @@ exposes a large number of resources with a small number of methods on each
 [Create][], [Update][], [Delete][]), or [custom methods][].
 
+A resource that supports a [Create][], [Update][], or [Delete][] must also
+support [Get][] to enable the user to validate the success of their operation
```
"user" --> "client"?

toumorokoshi replied:
a user may also want to verify? but I'm not strongly opinionated. 

do we want to normalize such nomenclature? I imagine "consumer" is another one.

toumorokoshi replied:
done.

toumorokoshi committed:
addressing feedback

use "validate the state of the resource", since that's actually what's being done here.
Not so much "operation completion", which should be detected via LRO or request completion.

loudej commented on aip/general/0121.md:
```diff
@@ -61,6 +61,9 @@ exposes a large number of resources with a small number of methods on each
 resource. The methods can be either the standard methods ([Get][], [List][],
 [Create][], [Update][], [Delete][]), or [custom methods][].
 
+A resource that supports a [Create][], [Update][], or [Delete][] **must** also
```
Is there any case where you would have a resource that can `List` only? That would require returning repeated resources, which each have a name in that collection by definition. Being a resource with a name could be enough to say `Get` is required?

At that point, you could say a resource **must** support `Get` and a resource collection **must** support `List` by definition - the mutable methods wouldn't need to be considered the reason.

bgrant0607 replied:
It sounds like the question is whether we can just mandate that all resources must support Get?

All of the examples I'm aware of that don't currently support Get (cloudsql users, tag bindings, bigquery reservation assignments, etc.) support List and at least one mutation.

toumorokoshi replied:
yeah, there's probably no good reason a *resource* API wouldn't support Get. It's a bit stronger of an assertion and it's harder to get approval the stronger the assertion, but I'll start there. 

Do we want to require List too? It seems better from a UX standpoint so you can retrieve a list of all resources you might want to operate on. But I could also imagine for security purposes you may not want to be able to enumerate all resources.

Although in the security case, you can control that ability with a permissions model.

bgrant0607 replied:
List: Good point. I wouldn't speculatively weaken the standard. *Must* implement List SGTM as a reasonable starting point.

toumorokoshi replied:
we could also go **must** for Get and **should** for List: i think the latter is more critical for uniform resources. 

But I'll change the PR for **must** to both and let's get some reviewers to chime in.

toumorokoshi replied:
updated, but I'll leave the thread open for discussion.

toumorokoshi committed:
addressing feedback

- Making Get and List required as requiring them for a mutable
  operation effectively makes a resource only valid if it supports
  these operations.

noahdietz commented on aip/general/0121.md:
```diff
@@ -61,6 +61,10 @@ exposes a large number of resources with a small number of methods on each
 resource. The methods can be either the standard methods ([Get][], [List][],
 [Create][], [Update][], [Delete][]), or [custom methods][].
 
+A resource **must** support at minimum [Get][] and [List][]: clients must be
```
This language conflicts with Singleton Resources that must not support List https://google.aip.dev/156 so I'm not sure we can enforce "must" for _List_. I think it has to be the following:
```suggestion
A resource **must** support at minimum [Get][] and **should** support [List][]: clients must be
```

bgrant0607 replied:
I disagree with that approach. If there's a valid exception, we should exempt that case only.

A resource must support Get and all non-singleton resources must support List.

noahdietz replied:
Fine, as long as the language doesn't conflict with supporting Singleton resources.

toumorokoshi replied:
added! ptal.

toumorokoshi committed:
addressing feedback

- Lists cannot exist in the case of singleton resources, so
  calling out that edge case specifically as not requiring List.

noahdietz reviewed (commented):
LGTM but please allow @shwoodard to Approve

shwoodard committed:
Merge branch 'master' into toum/require-get-method
